### ISSUE TITLE
compilation with debugging symbols very slow

### ISSUE THREAD

**Issue by llllllllll** [hanickadot/compile-time-regular-expressions#78]
First, thanks for this wonderful library. I am amazed at how fast it can match and isolate groups!

I was evaluating using ctre for some datetime parsing, like you have in the readme; however, I am finding that the compile times are extremely long when emitting debugging symbols. I have a file, which does have some other medium weight templates, but with only 3 regex literals in the file, the compilation time goes from less than a minute to 22 minutes. Without debugging symbols, it only takes 25 seconds, so I don't believe that it is due to slowness in the library itself.

I am compiling with gcc 9.1.0, and the 3 regular expressions I am using are:

```
(\\d{4})-(\\d{2})-(\\d{2})
(\\d{4})-(\\d{2})-(\\d{2})( |T)(\\d{2}):(\\d{2}):(\\d{2})
(\\d{4})-(\\d{2})-(\\d{2})( |T)(\\d{2}):(\\d{2}):(\\d{2})(\\.\\d{1,9})?
```

I would like to add many more formats, but I am waiting to implement that until I know what is going on with the compile times. This project is currently using C++ 17 and only gcc, so I am using the `_ctre` UDL. In testing, I have tried the C++20 cnttp version and noticed that it is a bit faster. I am not really able to reproduce this when I just compile the longest regex standalone, so I am wondering if there is a weird interaction with the existing code somewhere.

I would like to preserve debugging symbols if possible, and if needed I could try just moving all the ctre stuff to its own object without symbols, but I am wondering if there is something else to try? If it would help, I can see about getting the actual offending code to you on Monday.

**Comment by hanickadot** [comment]
The compilation time is really oddly big. I would like to see the offending code.

**Comment by osiewicz** [comment]
Hi all,
I believe I've created a [minimal working example](https://gist.github.com/osiewicz/674a68458dcf662573769781710ee036) for this issue. The code indeed takes over a minute to build even without debug symbol. I did not go the extra mile to wait for over 25 minutes on my vm, though I might do just that tommorow.

I believe I pinned down the potential offender: **forced inlining of code**. I've temporarily changed the definitions of [CTLL_FORCE_INLINE](https://github.com/hanickadot/compile-time-regular-expressions/blob/master/single-header/ctre.hpp#L465) and [CTRE_FORCE_INLINE](https://github.com/hanickadot/compile-time-regular-expressions/blob/master/single-header/ctre.hpp#L1336) & removed forced inlining. The benchmark results can be found [here](https://gist.github.com/osiewicz/995146990d86c269726368c8fc03d091). 

Of course it's easy to pinpoint the issue, though it does not mean that what I did is a correct way to handle long compile times - my comment is missing:
1. runtime benchmark of two versions.
2. precisely pinpointed cause of the issue - this is only something I can go after.

I suspect staying away from forced inlining is not an option.

Needless to say I might have done something incorrectly. Feel free to point out any issues.

**Comment by Archie3d** [comment]
Hi, I've spottedthis issue in latest MSVC 2019 (19.26.28806) now, except the debug compilation is fine, but in release (/Ox flag) the compilation never ends. Removing inlining indeed unblocks it.

**Comment by hanickadot** [comment]
@Archie3d does it happen for any pattern? if not, please provide your test case.

**Comment by Archie3d** [comment]
@hanickadot no, the shorter the pattern the faster it compiles. For this one it takes forever:
```cpp
static constexpr auto pattern = ctll::fixed_string { "^([a-z]+)://(([\\w\\-\\.]+)(:([\\w\\-\\?&=]+))?@)?([\\w\\-\\.]+)(:(\\d+)+)?(/[\\w\\.\\-/\\?\\(\\)&=%~]*)?$" };
```
though when I remove the trailing groups the compilation goes faster (than my patience wears out :)

Compilation flags (MSVC2019):
```
/EHsc /std:c++latest /bigobj /Ox
```
clang does not seem to suffer though.

**Comment by hanickadot** [comment]
That's a problem of MVSC and their constexpr evaluator :( It's not forever only too long. It's definitely a different issue than #78, please create a new one. 

This issue about MSVC creating a lot of debug symbols.

**Comment by nitronoid** [comment]
+1 for the patch provided by @osiewicz , it cut my build times down from 4 mins to 11 seconds with clang, and from 3 mins to 12 seconds with gcc.
I wonder if this could be merged with a macro to override the default inlining behaviour for better build times?
